cloud_blobstore/s3.py

```diff
@@ -10,9 +10,66 @@
     BlobStore,
     BlobStoreCredentialError,
     BlobStoreUnknownError,
+    PagedIter
 )
 
 
+class AWSPagedIter(PagedIter):
+    def __init__(
+            self,
+            bucket: str,
+            prefix: str=None,
+            delimiter: str=None,
+            marker: str=None,
+            token: str=None,
+            k_page_max: int=None
+    ) -> typing.Iterator[str]:
+        self.marker = marker
+        self.token = token
+        self.k_page_max = None
+
+        self.kwargs = {
+            'Bucket': bucket,
+        }
+
+        if prefix is not None:
+            self.kwargs['Prefix'] = prefix
+
+        if delimiter is not None:
+            self.kwargs['Delimiter'] = delimiter
+
+        if k_page_max is not None:
+            self.kwargs['MaxKeys'] = k_page_max
+
+    def get_api_response(self, next_token):
+        kwargs = self.kwargs.copy()
+
+        if next_token is not None:
+            kwargs['ContinuationToken'] = next_token
+    
+        resp = boto3.client('s3').list_objects_v2(
+            ** kwargs
+        )
+
+        return resp
+
+    def get_listing_from_response(self, resp):
+        if resp.get('Contents', None):
+            contents = resp['Contents']
+        else:
+            contents = list()
+
+        return [b['Key'] for b in contents]
```

Comment:
This obviously creates a new list. I don't think this is a huge problem since the list length is <= 1000. And the way the return value is consumed multiple times by the base dictates that this be a list.

However, I have the feeling that changing the calling method to use a single loop over the result set might actually make it simpler and more efficient. You could use `if key >= marker` to skip over previously seen elements in that loop.

Anyways, not a change request, just an observation.

Reply:
++

I would do what @hannes-ucsc suggested.

Reply:
Good point, I'll work it in.